packages/messaging/src/controllers/sw-controller.ts

```diff
@@ -139,34 +159,53 @@ export class SwController implements FirebaseMessaging, FirebaseService {
   }
 
   /**
-   * A handler for push events that shows notifications based on the content of
-   * the payload.
+   * A handler for push events that shows notifications based on the content of the payload.
    *
-   * The payload must be a JSON-encoded Object with a `notification` key. The
-   * value of the `notification` property will be used as the NotificationOptions
-   * object passed to showNotification. Additionally, the `title` property of the
-   * notification object will be used as the title.
+   * The payload must be a JSON-encoded Object with a `notification` key. The value of the
+   * `notification` property will be used as the NotificationOptions object passed to
+   * showNotification. Additionally, the `title` property of the notification object will be used as
+   * the title.
    *
-   * If there is no notification data in the payload then no notification will be
-   * shown.
+   * If there is no notification data in the payload then no notification will be shown.
    */
   async onPush(event: PushEvent): Promise<void> {
-    const payload = getMessagePayload(event);
-    if (!payload) {
+    const internalPayload = getMessagePayloadInternal(event);
+    if (!internalPayload) {
+      console.debug(
+        TAG +
+          'failed to get parse MessagePayload from the PushEvent. Skip handling the push.'
+      );
       return;
     }
 
+    // foreground handling: eventually passed to onMessage hook
     const clientList = await getClientList();
     if (hasVisibleClients(clientList)) {
-      // App in foreground. Send to page.
-      return sendMessageToWindowClients(clientList, payload);
+      return sendMessagePayloadInternalToWindows(clientList, internalPayload);
     }
 
-    const notificationDetails = getNotificationData(payload);
-    if (notificationDetails) {
-      await showNotification(notificationDetails);
-    } else if (this.bgMessageHandler) {
-      await this.bgMessageHandler(payload);
+    // background handling: display and pass to onBackgroundMessage hook
+    let isNotificationShown = false;
+    if (!!internalPayload.notification) {
+      await showNotification(wrapInternalPayload(internalPayload));
+      isNotificationShown = true;
+    }
+
+    if (
+      isNotificationShown === true &&
+      this.isOnBackgroundMessageUsed === false
```

Comment:
Can you put the comment for `isOnBackgroundMessageUsed` here as well? It will make it easier to understand without jumping around in the code.